ai/WFLY-19381-[EXPERIMENTAL]-Provide_a_galleon_feature_pack_to_facilitate_Genrative_AI_development.adoc

```diff
@@ -0,0 +1,264 @@
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [] Managed domain
+
+* [] OpenShift s2i
+
+* [] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+This feature should be available as an external galleon feature pack thus not being bound to a specific WildFly version or to the WildFly release cycle.
+The feature pack should enable a way to configure and provide resources to build RAG applications. 
+It should provide at least two kinds of:
+ * embedding models (aka models used to create embeddings): `dev.langchain4j.model.embedding.EmbeddingModel`
+ * embedding stores (aka places to store the computed embeddings): `dev.langchain4j.store.embedding.EmbeddingStore`
+ * content retrievers (aka retrievers of content to provide to the llm as part of the context based on the user query): `dev.langchain4j.rag.content.retriever.ContentRetriever`
+ * chat language models (aka a chat APi with the llm): `dev.langchain4j.model.chat.ChatLanguageModel`
+ Those resources should be exposed via CDI (and thus Weld) to the application using qualifier and type.
+ The less WildFly specific annotations are used the better so this feature should try to use annotations from librairies that are already used in WildFly like smallrye-common-annotations or annotations from LangChain4J.
+ We should provide layers to provision the server according to the needs.
+
+
+=== Nice-to-Have Requirements
+
+ * Provide annotations to be able to create AIServices using our configured resources.
+ * Replace the HTTP clients used so that we have only one that is supported for every API.
+ * Replace the JSON marshalling/umarshalling librairies so that we have only one (through RESTEasy) that is supported.
+ * Support for @Tool
+ * Adding support for ChatMemory
+ * Adding support for more APIs
+
+=== Non-Requirements
+// Use this section to explicitly discuss things that readers might think are required
+// but which are not required.
+
+=== Future Work
+// Use this section to discuss requirements that are not addressed by this proposal
+// but which may be addressed in later proposals.
+
+== Backwards Compatibility
+
+// Does this enhancement affect backwards compatibility with previously released
+// versions of WildFly?
+// Can the identified incompatibility be avoided?
+
+=== Default Configuration
+
+=== Importing Existing Configuration
+
+=== Deployments
+
+The required librairies should be added automatically on the deployment classpath. 
+
+=== Interoperability
+
+== Implementation Plan
+
+=== Embeddings models
+
+The extension should provide resources to define `dev.langchain4j.model.embedding.EmbeddingModel`.
+
+It should expose a simple `embedding-model` resource with the following attributes:
+
+* module: the jboss module containing the code of the model.
+* embedding-class: the name of the class to use to load the model.
+
```

Comment:
Do you think that we need to pass the "Vector Dimension" as a parameter as well? Like we have on https://docs.quarkiverse.io/quarkus-langchain4j/dev/in-process-embedding.html

Reply:
I don't think the vector dimension is a parameter, more a detail of the model itself

Reply:
I got it. I thought that could be a parameter because the Infinispan document store requires the dimension of the vector to be set, for example. :-)

Reply:
but that's a store while this is the part that computes the embeddings